### PR TITLE
Posts: Cleaner Post Controls component

### PR DESCRIPTION
**Pull request by Copons**
cc @aduth, @nb  (see: https://github.com/Automattic/wp-calypso/pull/7451#discussion_r74931463)

I rewrote the component in ES6 and moved the controls logic out of the `render` method for improved readability.
There is also a Copy control ready for the Copy Post functionality, so that this refactor could be merged right after the Copy Post merge.

### CODE REVIEW

File: client/my-sites/posts/post-controls.jsx [hunk 1]
```diff
@@ -1,226 +1,242 @@
...
+	}
+
+	buildControls( controls ) {
+		return controls.map( ( control, i ) => {
 			return (
-				<li key={ itemKey }>
-					<a href={ item.href } className={ item.className } onClick={ item.onClick } target={ item.target ? item.target : null }>
-						<Gridicon icon={ item.icon } size={ 18 } />
-						<span>{ item.text }</span>
+				<li key={ `controls-${this.props.post.ID}-${i}` }>
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: I was surprised to find that this didn't warn on the lack of spaces between curly braces. Looks like we haven't opted into [`template-curly-spacing`](http://eslint.org/docs/rules/template-curly-spacing), which I've created a pull request for at https://github.com/Automattic/eslint-config-wpcalypso/pull/1 .

File: client/my-sites/posts/post-controls.jsx [hunk 2]
```diff
@@ -1,226 +1,242 @@
...
+		onShowMore: PropTypes.func.isRequired,
+		onHideMore: PropTypes.func.isRequired,
+		onPublish: PropTypes.func,
+		onTrash: PropTypes.func,
+		onDelete: PropTypes.func,
+		onRestore: PropTypes.func,
+		translate: PropTypes.func,
+	};
+
+	constructor() {
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: You don't need a `constructor` if you're not doing anything aside from `super` in it (the inherited `constructor` will be called by default).

File: client/my-sites/posts/post-controls.jsx [hunk 3]
```diff
@@ -1,226 +1,242 @@
 /**
  * External dependencies
  */
-var React = require( 'react' ),
-	PureRenderMixin = require( 'react-pure-render/mixin' ),
-	url = require( 'url' );
+import React, { PropTypes } from 'react';
+import PureComponent from 'react-pure-render/component';
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: As of [React 15.3.0](https://github.com/facebook/react/blob/master/CHANGELOG.md#1530-july-29-2016), you can now `import { PureComponent } from 'react';` . We'll plan to convert existing `react-pure-render` to the available equivalents and remove it as a dependency altogether.

File: client/my-sites/posts/post-controls.jsx (hunk 1 above)
Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: Ah!
It looks like I'll have to do quite some fixes all around, as the lack of warning made me think that it was intended! :laughing:

File: client/my-sites/posts/post-controls.jsx [hunk 4]
```diff
@@ -1,226 +1,237 @@
...
-	},
+		} );
+	}
 
-	render: function() {
-		var post = this.props.post,
-			availableControls = [],
-			extraControls = [],
-			parsed, previewURL, statsURL, postControlsClass, mainControls, moreControls, i, moreControlsSpliceIndex;
+	setControls() {
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: Since this isn't really setting anything, do you think `getControls` might be a more appropriate name?

File: client/my-sites/posts/post-controls.jsx [hunk 5]
```diff
@@ -1,226 +1,237 @@
...
-					<ul className="post-controls__pane post-controls__main-options">
-						{ mainControls }
-					</ul>
-				</div>
-			);
-		} else {
-			mainControls = this.buildControls( availableControls );
+	render() {
+		const controls = this.setControls();
+		let postControlsClass = 'post-controls';
```
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: The [`classnames` module](https://github.com/JedWatson/classnames) would be a nicer way to construct the class name I think. Something like:

```js
const className = classNames( 'post-controls', {
	'post-controls--desk-nomore': controls.more.length <= 2
} );
```

We should also seek to update to our [classname naming guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#csssass-coding-guidelines), though I think that could be saved for a separate pull request.

File: client/my-sites/posts/post-controls.jsx (hunk 4 above)
Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: I honestly can't remember the reasoning behind the change from `buildControls` to `setControls`.
As you comment is definitely correct, I guess I could just change it back to `buildControls` as I find it more appropriate than `getControls`, considering the method actually _creates_ the controls arrays.

File: client/my-sites/posts/post-controls.jsx (hunk 5 above)
Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: I've noticed that `post-controls--desk-nomore` isn't used anywhere else in Calypso.
So I thought I could just remove it and update all the other classes to `posts__post-controls.modifiers`, but I'm wary of letting this branch grow too much.

**Comment by Copons** [comment]
@aduth I can easily remove the Copy button and add it in the Copy Post branch once this PR is merged.

File: client/my-sites/posts/post-controls.jsx (hunk 5 above)
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: >I've noticed that `post-controls--desk-nomore` isn't used anywhere else in Calypso.
So I thought I could just remove it

Hmm, yeah, if it's not in the codebase, probably safe to remove. Based on the name, perhaps something from desktop app? I don't see anything in [that codebase](https://github.com/Automattic/wp-desktop) either (cc @sendhil in case this rings any bells).

File: client/my-sites/posts/post-controls.jsx (hunk 4 above)
Comment (aduth) [inline client/my-sites/posts/post-controls.jsx]: There's still `buildControls` above which appears to be a helper function for generating the elements based on the object structure. I'm not sure what the best naming might be, maybe `getAvailableControls` and `getControlElements`?

Comment (Copons) [inline client/my-sites/posts/post-controls.jsx]: Ouch, I totally missed that!
I'm gonna think about it (which likely means: "I'm going with your suggestions" 😆 ) while we wait for the wp-desktop confirmation about the class name.

File: client/my-sites/posts/post-controls.jsx (hunk 5 above)
Comment (sendhil) [inline client/my-sites/posts/post-controls.jsx]: @aduth - I don't think it is but could you all do a sanity check with this PR on the desktop repo?

**Comment by Copons** [comment]
Moved to https://github.com/Automattic/wp-calypso/pull/10746
